src/Storages/MergeTree/MergeTreeDataPartCloner.cpp

```diff
@@ -0,0 +1,395 @@
+    MergeTreeData * merge_tree_data,
+    const MergeTreeData::DataPartPtr & src_part
+)
+{
+    for (const DiskPtr & disk : merge_tree_data->getStoragePolicy()->getDisks())
+    {
+        if (disk->getName() == src_part->getDataPartStorage().getDiskName())
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+std::pair<MergeTreeData::MutableDataPartPtr, scope_guard> MergeTreeDataPartCloner::cloneSourcePart(
+    MergeTreeData * merge_tree_data,
+    const MergeTreeData::DataPartPtr & src_part,
+    const StorageMetadataPtr & metadata_snapshot,
+    const MergeTreePartInfo & dst_part_info,
+    const String & tmp_part_prefix,
+    const ReadSettings & read_settings,
+    const WriteSettings & write_settings,
+    const DB::IDataPartStorage::ClonePartParams & params
+)
+{
+    const auto dst_part_name = src_part->getNewName(dst_part_info);
+
+    const auto tmp_dst_part_name = tmp_part_prefix + dst_part_name;
+
+    auto temporary_directory_lock = merge_tree_data->getTemporaryPartDirectoryHolder(tmp_dst_part_name);
+
+    reserveSpaceOnDisk(src_part);
+
+    scope_guard src_flushed_tmp_dir_lock;
+    MergeTreeData::MutableDataPartPtr src_flushed_tmp_part;
+
+    auto src_part_storage = flushPartStorageToDiskIfInMemory(
+        merge_tree_data,
+        src_part,
+        metadata_snapshot,
+        tmp_part_prefix,
+        tmp_dst_part_name,
+        src_flushed_tmp_dir_lock,
+        src_flushed_tmp_part
+    );
+
+    auto dst_part_storage = hardlinkAllFiles(
+        merge_tree_data,
+        read_settings,
+        write_settings,
+        src_part_storage,
+        tmp_dst_part_name,
+        params
+    );
+
+    if (params.metadata_version_to_write.has_value())
+    {
+        chassert(!params.keep_metadata_version);
+        auto out_metadata = dst_part_storage->writeFile(IMergeTreeDataPart::METADATA_VERSION_FILE_NAME, 4096, merge_tree_data->getContext()->getWriteSettings());
+        writeText(metadata_snapshot->getMetadataVersion(), *out_metadata);
+        out_metadata->finalize();
+        if (merge_tree_data->getSettings()->fsync_after_insert)
+            out_metadata->sync();
+    }
+
+    LOG_DEBUG(log, "Clone {} part {} to {}{}",
+              src_flushed_tmp_part ? "flushed" : "",
+              src_part_storage->getFullPath(),
+              std::string(fs::path(dst_part_storage->getFullRootPath()) / tmp_dst_part_name),
+              false);
+
+
+    auto part = MergeTreeDataPartBuilder(*merge_tree_data, dst_part_name, dst_part_storage)
+                    .withPartFormatFromDisk()
+                    .build();
+
+    return std::make_pair(part, std::move(temporary_directory_lock));
+}
+
+void MergeTreeDataPartCloner::reserveSpaceOnDisk(const MergeTreeData::DataPartPtr & src_part)
+{
+    src_part->getDataPartStorage().reserve(src_part->getBytesOnDisk());
+}
+
+DataPartStoragePtr MergeTreeDataPartCloner::flushPartStorageToDiskIfInMemory(
```

Comment:
InMemory parts are deprecated, we don't need to support this

Reply:
There is some logic to handle this in current code: https://github.com/ClickHouse/ClickHouse/pull/39507/files#diff-b7d33a7abda5f3365d5e4fb58ca98b756f6401372b2e2d5eb449a2bd3444fb5eL6984.

Can I safely remove this?

Reply:
Let's leave it for compatibility (I didn't like adding new code for the deprecated feature, but it's okay since the copy-paste is removed)